store/src/java/com/zimbra/cs/mailbox/MailSender.java

```diff
@@ -1478,4 +1479,45 @@ public static void unregisterPreSendMailListener(PreSendMailListener listener) {
             }
         }
     }
+
+    /**
+     * provide the Hash for Message-Verification header field.
+     * @param id
+     * @param date
+     * @param from
+     * @throws ServiceException
+     *
+     */
+    private static String getMessageVerificationHeaderValue(String id, Date date, String from) throws MessagingException, ServiceException {
+        String guid = (id + date + from);
```

Comment:
Is it secure? Don't we need to add any system-wide (= globalConfig) secret value?

https://www.ietf.org/archive/id/draft-leiba-morg-message-recall-00.html#anchor4
The actual GUID (the pre-image of the hash) is kept secret by the sending side.

Reply:
**Yes you are right, we need to add some secret to the sha256 hash.**

Otherwise a user could both spoof the from and sha256 header and still recall any message with certain message ID. Which is what we are trying to avoid.

In addition when we recall a message, the authenticated user account name and or alias should be validated against the from header. If that matches **_and_** the sha256 hash is correct we can proceed and recall.

Reply:
Done

Reply:
Implemented the secret key at global level